### ISSUE TITLE
Rule idea: warn about unsafe usage of `$` in regular expressions

### ISSUE THREAD

**Issue by bustbr** [astral-sh/ruff#15017]
In Python's implementation of `re` the `$` does not only match the end of the line, but also before any line break, even without multiline mode.
There's [an article by the OpenSSF about this issue](https://best.openssf.org/Correctly-Using-Regular-Expressions) suggesting to use `\Z` instead, or prefer the `fullmatch` function.

An example of the issue:
```
>>> re.search(r'cat$', 'cat\n')
<re.Match object; span=(0, 3), match='cat'>
```

Using `\Z` fixes this:
```
>>> re.search(r'cat\Z', 'cat\n')  # no match
>>> re.search(r'cat\Z', 'cat')
<re.Match object; span=(0, 3), match='cat'>
```

Or using `fullmatch`:
```
>>> re.fullmatch('cat', 'cat\n')  # no match
>>> re.fullmatch('cat', 'cat')
<re.Match object; span=(0, 3), match='cat'>
```

Should Ruff warn about using `$` with `search`?

**Comment by bustbr** [comment]
> > In Python's implementation of `re` the `$` does not only match the end of the line, but also before any line break, even without multiline mode.
> 
> This is incorrect. In non-multiline mode, `$` matches either the end of the entire input or the position right before the trailing newline (only `\n`, not `\r\n`). It is similar to PCRE's `\Z`, whereas Python's `\Z` resembles PCRE's `\z`:

Thank you for the clarification.
Based on your example here's a series of calls to highlight the specific behavior of `$`:

```python-console
>>> re.search(r'$', 'foo\nbar')
<re.Match object; span=(7, 7), match=''>
>>> re.search(r'$', 'foo\nbar\n')
<re.Match object; span=(7, 7), match=''>    # still matches at pos 7, before the final new line
>>> re.search(r'$', 'foo\nbar\n\n')
<re.Match object; span=(8, 8), match=''>    # matches at pos 8, before the final new line

>>> re.search(r'\Z', 'foo\nbar')
<re.Match object; span=(7, 7), match=''>
>>> re.search(r'\Z', 'foo\nbar\n')
<re.Match object; span=(8, 8), match=''>
>>> re.search(r'\Z', 'foo\nbar\n\n')
<re.Match object; span=(9, 9), match=''>
```
